src/UsgsAstroLsSensorModel.cpp

```diff
@@ -1666,6 +1652,138 @@ double UsgsAstroLsSensorModel::getValue(
+   losIsis[2] = -m_focal * (1.0 - getValue(15, adj) / m_halfSwath);
+   double isisMag = sqrt(losIsis[0] * losIsis[0]
+      + losIsis[1] * losIsis[1]
+      + losIsis[2] * losIsis[2]);
+   losIsis[0] /= isisMag;
+   losIsis[1] /= isisMag;
+   losIsis[2] /= isisMag;
+};
+
+
+// Given a time and a flag to indicate whether the a->b or b->a rotation should be calculated
+// uses the quaternions in the m_quaternions member to calclate a rotation matrix. 
+void UsgsAstroLsSensorModel::calculateRotationMatrixFromQuaternions(const double& time, const bool& invert, double rotationMatrix[9]) const {
+  int nOrder = 8;
+  if (m_platformFlag == 0)
+     nOrder = 4;
+  int nOrderQuat = nOrder;
+  if (m_numQuaternions < 6 && nOrder == 8)
+     nOrderQuat = 4;
+  double q[4];
+  lagrangeInterp(
+     m_numQuaternions, &m_quaternions[0], m_t0Quat, m_dtQuat,
+     time, 4, nOrderQuat, q);
+  double norm = sqrt(q[0] * q[0] + q[1] * q[1] + q[2] * q[2] + q[3] * q[3]);
+
+  if (!invert) {
+    q[0] /= norm; 
+    q[1] /= norm;
+    q[2] /= norm;
+    q[3] /= norm;
+  }
+  else {
+    q[0] /= -norm; 
+    q[1] /= -norm;
+    q[2] /= -norm;
+    q[3] /= norm;
+  }
+
+  rotationMatrix[0] = q[0] * q[0] - q[1] * q[1] - q[2] * q[2] + q[3] * q[3];
+  rotationMatrix[1] = 2 * (q[0] * q[1] - q[2] * q[3]);
+  rotationMatrix[2] = 2 * (q[0] * q[2] + q[1] * q[3]);
+  rotationMatrix[3] = 2 * (q[0] * q[1] + q[2] * q[3]);
+  rotationMatrix[4] = -q[0] * q[0] + q[1] * q[1] - q[2] * q[2] + q[3] * q[3];
+  rotationMatrix[5] = 2 * (q[1] * q[2] - q[0] * q[3]);
+  rotationMatrix[6] = 2 * (q[0] * q[2] - q[1] * q[3]);
+  rotationMatrix[7] = 2 * (q[1] * q[2] + q[0] * q[3]);
+  rotationMatrix[8] = -q[0] * q[0] - q[1] * q[1] + q[2] * q[2] + q[3] * q[3];
+};
+
+
+void UsgsAstroLsSensorModel::calculateAttitudeCorrection(const double& time, const std::vector<double>& adj, double attCorr[9]) const {
+  double aTime = time - m_t0Quat;
+    double euler[3];
+    double nTime = aTime / m_halfTime;
+    double nTime2 = nTime * nTime;
+    euler[0] =
+       (getValue(6, adj) + getValue(9, adj)* nTime + getValue(12, adj)* nTime2) / m_flyingHeight;
+    euler[1] =
+       (getValue(7, adj) + getValue(10, adj)* nTime + getValue(13, adj)* nTime2) / m_flyingHeight;
+    euler[2] =
+       (getValue(8, adj) + getValue(11, adj)* nTime + getValue(14, adj)* nTime2) / m_halfSwath;
+    double cos_a = cos(euler[0]);
+    double sin_a = sin(euler[0]);
+    double cos_b = cos(euler[1]);
+    double sin_b = sin(euler[1]);
+    double cos_c = cos(euler[2]);
+    double sin_c = sin(euler[2]);
+
+    attCorr[0] = cos_b * cos_c;
+    attCorr[1] = -cos_a * sin_c + sin_a * sin_b * cos_c;
+    attCorr[2] = sin_a * sin_c + cos_a * sin_b * cos_c;
+    attCorr[3] = cos_b * sin_c;
+    attCorr[4] = cos_a * cos_c + sin_a * sin_b * sin_c;
+    attCorr[5] = -sin_a * cos_c + cos_a * sin_b * sin_c;
+    attCorr[6] = -sin_b;
+    attCorr[7] = sin_a * cos_b;
+    attCorr[8] = cos_a * cos_b;

```

Comment:
This is a rotation matrix calculated from euler angles. It would be nice to have a function similar to the one for quaternions that computes a rotation matrix from euler angles.

Reply:
Sure, this sounds like a good idea, and will do.